pandas/_libs/lib.pyx

```diff
@@ -2359,3 +2356,106 @@ def fast_multiget(dict mapping, ndarray keys, default=np.nan):
             output[i] = default
 
     return maybe_convert_objects(output)
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef inline void convert_and_set_item(object item, Py_ssize_t index,
+                                      object[:] result,
+                                      bint keep_trivial_numbers):
+    cdef:
+        bint do_convert = 1
+        double double_item
+
+    if keep_trivial_numbers:
+        if isinstance(item, int) and Py_SIZE(item) < 2:
+            if <int>item == 0:
+                do_convert = 0
+        elif isinstance(item, float):
+            double_item = item
+            if double_item == 0.0 or double_item != double_item:
+                do_convert = 0
+
+    if do_convert and not isinstance(item, (str, bytes)):
+        item = PyObject_Str(item)
+
+    result[index] = item
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef inline void put_object_as_unicode(list lst, Py_ssize_t idx,
+                                       object item):
+    if not isinstance(item, str):
+        item = PyObject_Str(item)
+    lst[idx] = item
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cpdef object _concat_date_cols(tuple date_cols,
+                               object keep_trivial_numbers=False):
+    cdef:
+        bint keep_numbers
+        Py_ssize_t sequence_size, i, j
+        Py_ssize_t array_size, min_size = 0
+        object[:] result_view
+
+        flatiter it
+        int all_numpy = 1
+        cnp.ndarray[object] iters
+        object[::1] iters_view
+        list list_to_join
+
+    keep_numbers = keep_trivial_numbers
+    sequence_size = len(date_cols)
+
+    if sequence_size == 0:
+        result = np.zeros(0, dtype=object)
+    elif sequence_size == 1:
+        array = date_cols[0]
+        array_size = len(array)
+        result = np.zeros(array_size, dtype=object)
+        result_view = result
+        if PyArray_Check(array):
+            it = <flatiter>PyArray_IterNew(array)
```

Comment:
why are you not simply writing loops here? I find this *really* hard to believe

Reply:
I check this. If write here something like that:
```python
for i, item in enumerate(array):
    convert_and_set_item(item, i, result_view, keep_trivial_numbers)
```
then for numpy array case we will get runtime, that x4 slower relative explicit use PyArray_Iter*